content/blog/performant-code/index.Rmd

```diff
@@ -0,0 +1,526 @@
+
+    - If not, could it? i.e. if the elements of the join are not indices of a loop, or are indices that could easily be subsetted on after the fact, do the join just once higher up in the call stack.
+
+- Am I using the complete outputted join result or just a portion? If I end up only making use of column names, or values in one column, or pairings between two columns, I may be able to instead use `$.tbl_df` or `[.tbl_df`.
+
+
+As an example, imagine we have another tibble that tells us additional information about the `make_model`s that I drove in high school:
+
+```{r simons-car}
+simons_cars <- 
+  tibble(
+    make_model = "Honda Civic",
+    vibey = "Yes"
+  )
+
+simons_cars
+```
+
+I _could_ use a join to subset down to only the cars I drove in high school and append this additional information like so:
+
+```{r right-join}
+right_join(mtcars_tbl, simons_cars, "make_model")
+```
+
+Another way to express this, though, is to keep rows that match the `make_model` in `simons_cars` and then add that `vibey` column on after:
+
+```{r supplement-simons-cars}
+supplement_simons_cars <- function() {
+  res <-   
+    vec_slice(
+      mtcars_tbl, 
+      mtcars_tbl$make_model %in% simons_cars$make_model
+    )
+  
+  res$vibey <- simons_cars$vibey
+  
+  res
+}
+
+supplement_simons_cars()
+```
+
+This is indeed quite a bit faster:
+
+```{r bm-join}
+bench::mark(
+  right_join = right_join(mtcars_tbl, simons_cars, "make_model"),
+  manual = supplement_simons_cars()
+) %>%
+  select(expression, median)
+```
+
+At the same time, if this problem were even a little bit more complex, e.g. if I drove more than one car in high school or wanted to keep the remaining rows and assert that all other cars were not vibey, then expressing this join with more bare-bones operations quickly becomes less readable and more error-prone. In those cases, too, joins in dplyr have a relatively small amount of overhead when compared to the vctrs backends underlying them. So, optimize carefully!
+
+Some examples of writing out joins in tidymodels packages: [tidymodels/parsnip#932](https://github.com/tidymodels/parsnip/pull/932), [tidymodels/parsnip#931](https://github.com/tidymodels/parsnip/pull/931), [tidymodels/parsnip#921](https://github.com/tidymodels/parsnip/pull/921), and [tidymodels/recipes#1121](https://github.com/tidymodels/recipes/pull/1121).
+
+### `nest()`
+
+`nest()`s are subject to similar considerations as joins. When they allow for expressive or principled user interfaces, use them, but manipulate them sparingly in backends. Writing out `nest()` calls _can_ result in substantial speedups, though, and the process is not quite as gnarly as writing out a join. For code that relies on `nest()`s and sees heavy traffic, rewriting with vctrs may be worth the effort.
+
+For example, consider nesting `mtcars_tbl` by `cyl`:
+
+```{r nest}
+nest(mtcars_tbl, .by = cyl)
+```
+
+For some basic nests, like nesting by values in one column, `vec_split()` can do the trick.
```

Comment:
`vec_split()` takes data frames so you can nest by more than one column too. I think it would actually be worth showing an example of this where you'd do something like:

```r
vec_cbind(res$key, data_frame(data = res$val))
```